### PR TITLE
cql: check if the function is builtin when granting permissisons

### PR DESCRIPTION
**Pull request by wmitros**
Currently, when granting a permission on a funciton resource, we only check if the function exists, regardless of whether it's a user or a builtin function. We should not support altering permissions on builtin functions, so this patch adds a check for confirming that the found function is not builtin.

Additionally, adjust an error exception thrown when trying to alter a permission that does not apply on a given resource

### CODE REVIEW

**Comment by wmitros** [comment]
Found by @bhalevy in [https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-debug/164/te[…]plit007___test_disallow_grant_revoke_on_builtin_functions/](https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-debug/164/testReport/auth_roles_test/TestAuthRoles/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split007___test_disallow_grant_revoke_on_builtin_functions/)

File: service/client_state.cc [hunk 1]
```diff
@@ -254,6 +257,20 @@ future<> service::client_state::ensure_exists(const auth::resource& r) const {
     });
 }
 
+void service::client_state::ensure_alterable_permisssions(const auth::resource& r) const {
```
Comment (nyh) [inline service/client_state.cc:260]: I don't see why this new code should be in client_state.cc - as far as I understand (but correct me if I'm wrong), it doesn't have anything to do with the client's permissions - it's just an intrinsic feature of the function name. So perhaps you can put it in permission_altering_statement.cc?

File: service/client_state.cc [hunk 2]
```diff
@@ -254,6 +257,20 @@ future<> service::client_state::ensure_exists(const auth::resource& r) const {
...
+    // We assume we have already checked that the resource exists.
+    if (r.kind() == auth::resource_kind::functions) {
+        auth::functions_resource_view v(r);
+        const auto keyspace = v.keyspace();
+        const auto function_signature = v.function_signature();
```
Comment (nyh) [inline service/client_state.cc:265]: silly nitpick: I think it doesn't help to add these shortcuts variables, each of them only used in one place.

**Comment by wmitros** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/34/

Seems related to #12629, I'll try rebuilding

**Comment by bhalevy** [comment]
Please verify the whole auth_roles_test.py dtest.
There are other regression that may be related to 843a5dfc15197eaadb4e46e7252f7b323373b523
like https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/218/testReport/auth_roles_test/TestAuthRoles/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split000___test_disallow_grant_execute_on_non_function_resources/
```
cassandra.InvalidRequest: Error from server: code=2200 [Invalid query] message="Resource <all keyspaces> does not support any of the requested permissions."
```

**Comment by wmitros** [comment]
@bhalevy I was considering changing the dtest case that you mentioned, because the only thing that was failing there was the type of the raised exception, but the change ended up so simple on the scylla side that I just fixed this here. There should be no more errors in `auth_roles_test.py` now

**Comment by bhalevy** [comment]
> @bhalevy I was considering changing the dtest case that you mentioned because the only thing that was failing there was the type of the raised exception, but the change ended up so simple on the scylla side that I just fixed this here. There should be no more errors in `auth_roles_test.py` now

Awesome, thanks!

File: cql3/statements/permission_altering_statement.cc [hunk 3]
```diff
@@ -12,6 +12,9 @@
 
 #include "auth/service.hh"
 #include "permission_altering_statement.hh"
+#include "cql3/functions/functions.hh"
+#include "cql3/functions/user_aggregate.hh"
```
Comment (avikivity) [inline cql3/statements/permission_altering_statement.cc:16]: Why "service:" in changelog? this is cql3.

File: cql3/statements/permission_altering_statement.cc [hunk 4]
```diff
@@ -45,6 +48,17 @@ future<> cql3::statements::permission_altering_statement::check_access(query_pro
     maybe_correct_resource(_resource, state, qp);
 
     return state.ensure_exists(_resource).then([this, &state] {
+        if (_resource.kind() == auth::resource_kind::functions) {
+            // Even if the function exists, it may be a builtin function, in which case we disallow altering permissions on it.
+            auth::functions_resource_view v(_resource);
+            if (v.function_signature()) {
+                auto [name, function_args] = auth::decode_signature(*v.function_signature());
+                auto fun = cql3::functions::functions::find(db::functions::function_name{sstring(*v.keyspace()), name}, function_args).get();
+                if (!dynamic_cast<cql3::functions::user_function*>(fun) && !dynamic_cast<cql3::functions::user_aggregate*>(fun)) {
```
Comment (avikivity) [inline cql3/statements/permission_altering_statement.cc:57]: There is an is_native() method.

What happens if there is no signature?

Comment (avikivity) [reply cql3/statements/permission_altering_statement.cc]: Please add a comment then.

Comment (wmitros) [reply cql3/statements/permission_altering_statement.cc]: Ok

**Comment by wmitros** [comment]
@nyh @avikivity I addressed both of your comments, maybe we can try merging this

**Pull request merged**